### PR TITLE
feat(ws1): add NativeKVCacheAttnOp pure-PyTorch KV-cache attention reference

### PR DESCRIPTION
**Pull request by maxiaosong1124**
## ⚠️ Depends on #188 (attention) — do not merge first

`NativeKVCacheAttnOp` **delegates to the same `NativeAttentionOp` reduction** added in #188,
so prefill and decode share one numerical path. This branch is stacked on top of #188's
`feat/ws1-attention-pytorch-op`.

**Because #188 is not merged yet, the diff below currently also contains #188's attention
files.** For this review, please look only at the kv_cache-specific changes:

- `rl_engine/kernels/ops/pytorch/attention/kv_cache.py`
- `tests/test_kv_cache_attention.py`
- the `kv_cache_attention` entries in `rl_engine/kernels/registry.py`
- the kv_cache CI step in `.github/workflows/ci.yml`

Once #188 merges to `main`, I'll rebase onto `main` and the diff becomes lean (attention-free).
Please **do not merge this before #188**.

## Summary

Pure-PyTorch fp32 ground-truth reference for **KV-cache (incremental/decode) attention**
(WS1 / ISSUE #108). It concatenates cached and new K/V along the sequence axis and then **calls
the exact same `NativeAttentionOp`** used for full-sequence prefill — re-implementing softmax
here would defeat the purpose. This guarantees decode (rollout) and prefill (training) take an
identical reduction, which is what keeps rollout↔training numerically consistent.

Shapes follow Qwen3-8B (synthetic tensors, no weights downloaded): 32 Q / 8 KV heads
(GQA g=4) / head_dim=128, scale = 1/√128.

## What this adds

- **`NativeKVCacheAttnOp`** (`pytorch/attention/kv_cache.py`):
  `forward` / `forward_fp32(q, k_cache, v_cache, k_new, v_new, *, causal=True, scale=None,
  key_padding_mask=None)` per the frozen contract (INTERFACES §9). `cat([cache, new], dim=2)`
  → delegates to `NativeAttentionOp`. Returns attention output only; cache update is out of
  scope for the numerical contract.
- **Registry**: `PYTORCH_NATIVE_KV_CACHE_ATTN` enum + `kv_cache_attention` priority-map entries
  (cpu / cuda / rocm).
- **`tests/test_kv_cache_attention.py`** — 14 CPU-safe tests (+ GPU/large markers).
- **CI**: `pytest tests/test_kv_cache_attention.py -k "not large and not gpu"`.

## Test coverage

- **Delegation correctness** — `forward` / `forward_fp32` equal `NativeAttentionOp` on the
  concatenated K/V.
- **Prefill↔decode consistency** — stepwise decode matches a single full-sequence prefill.
- **Cache split-point invariance** — same total KV split at different cache/new boundaries
  gives the same result.
- **Axis-A** — batch slice invariance (`torch.equal`).
- **key_padding_mask** — padded keys get zero weight; near-equality `atol=2e-6` (differing
  softmax reduction widths, same rationale as #188's padding test, with headroom over the
  ~1.3e-6 platform-sensitive drift).
- **Axis-B** — bf16 / fp16 forward tracks fp32 within per-dtype peak-relative tolerance.
- Empty-cache == plain attention, closed-form uniform-attention decode, output shape follows
  `q`, GQA divisibility guard, input purity (no mutation), registry dispatch.

## Local run

\`\`\`
python -m pytest tests/test_kv_cache_attention.py -v -k "not large and not gpu"
# 14 passed
\`\`\`

## Notes
- Pure function: no randomness, no in-place, no device moves; dtype/device follow inputs.
- Mask convention: `True = valid / keep`, `False = padding` (consistent with the rest of the repo).

## Summary by CodeRabbit

* **New Features**
  * Added support for KV-cache attention with a native PyTorch implementation and registry routing.
  * Expanded CI to run additional CPU-safe attention validation checks.

* **Bug Fixes**
  * Enforced stricter input alignment for decode-time attention.
  * Improved handling for cached, masked, and incremental attention cases.

* **Tests**
  * Added comprehensive coverage for KV-cache attention correctness, precision, and error handling.

### CODE REVIEW

**Comment by Flink-ddd** [comment]
@maxiaosong1124 please resolve the code conflicts first, Thank. cc @KJLdefeated PTAL

**Comment by maxiaosong1124** [comment]
> @maxiaosong1124 please resolve the code conflicts first, Thank. cc @KJLdefeated PTAL

@Flink-ddd Resolved!

**Review by Flink-ddd — APPROVED** [review]
LGTM, Thanks. also cc @KJLdefeated PTAL.

**Review by KJLdefeated — APPROVED** [review]
LGTM. Thx!
